include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,252 @@
+  tracking_resource_adaptor& operator=(tracking_resource_adaptor const&) = delete;
+  tracking_resource_adaptor& operator=(tracking_resource_adaptor&&) = default;
+
+  /**
+   * @brief Return pointer to the upstream resource.
+   *
+   * @return Upstream* Pointer to the upstream resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @brief Checks whether the upstream resource supports streams.
+   *
+   * @return true The upstream resource supports streams
+   * @return false The upstream resource does not support streams.
+   */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+  /**
+   * @brief Print any outstanding allocations to debug log
+   *
+   */
+  void print_outstanding_allocations() const
+  {
+    lock_t lock(mtx);
+    std::ostringstream oss;
+    for (auto const& al : allocations) {
+      oss << std::endl
+          << "Allocation " << al.first << " of " << al.second.allocation_size
+          << "bytes with callstack:" << std::endl;
+      std::unique_ptr<char*, decltype(&::free)> strings(
+        backtrace_symbols(al.second.stack_ptrs.data(), al.second.stack_ptrs.size()), &::free);
+      if (strings.get() == nullptr) {
+        oss << "But no stack trace could be found!" << std::endl;
+      } else {
+        ///@todo: support for demangling of C++ symbol names
+        for (int i = 0; i < al.second.stack_ptrs.size(); ++i) {
+          oss << "#" << i << " in " << strings.get()[i] << std::endl;
+        }
+      }
+    }
+
+    RMM_LOG_DEBUG("Outstanding Allocations: {}", oss.str());
+  }
+
+  /**
+   * @brief Get the number of outstanding allocations
+   *
+   * @return std::size_t number of allocations still outstanding
+   */
+  std::size_t get_num_outstanding_allocations() const { return allocations.size(); };
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = [&]() {
+      try {
+        return upstream_->allocate(bytes, stream);
+      } catch (std::exception const& e) {
+        RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE {}]",
```

Comment:
Seems like the upstream resource should log the failure, and if it does, we will get duplicate logging.